derive/lib.rs

```diff
@@ -0,0 +1,116 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#![cfg_attr(not(test), feature(proc_macro, proc_macro_lib))]
+
+#[cfg(not(test))] extern crate proc_macro;
+#[macro_use] extern crate quote;
+extern crate syn;
+
+#[cfg(not(test))]
+#[proc_macro_derive(HeapSizeOf)]
+pub fn expand_token_stream(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    expand_string(&input.to_string()).parse().unwrap()
+}
+
+fn expand_string(input: &str) -> String {
```

Comment:
I see the previously existing implementation used `combine_substructure`. Take a look at [`synstructure`](https://github.com/mystor/synstructure) for similar functionality against syn. It is used by the `gc` crate and `abomonation_derive`.